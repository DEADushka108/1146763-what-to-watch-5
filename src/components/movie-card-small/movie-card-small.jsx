import React, {useCallback, useEffect, useRef, useState} from 'react';
import {Link} from 'react-router-dom';
import PropTypes from 'prop-types';
import {AppRoute} from '../../utils/const';
import {movieDetails} from '../../types/types';
import {redirectToRoute} from '../../store/redirect/redirect.js';
import {connect} from 'react-redux';

const VIDEO_PLAY_TIMEOUT = 1000;

const MovieCardSmall = (props) => {
  const {movie, onCardClick} = props;
  const {id, title, poster, previewSrc} = movie;
  const video = useRef();
  const [isPlay, setPlayStatus] = useState(false);
  let timeout;

  const handleMouseEnter = useCallback(() => {
    video.current.load();
    setPlayStatus(true);
    timeout = setTimeout(() => {
      if (!isPlay && video.current) {
        video.current.muted = true;
        video.current.play();
      }
    }, VIDEO_PLAY_TIMEOUT);
  }, [video]);

  const handleMouseLeave = useCallback(() => {
    setPlayStatus(false);
    clearTimeout(timeout);
    video.current.load();
  }, [video]);

  useEffect(() => {
    if (!video.current) {
      setPlayStatus(false);
      return;
    }
    return;
  }, [video.current]);

  return <article className="small-movie-card catalog__movies-card" >
    <div className="small-movie-card__image" onClick={() => {
      onCardClick(`${AppRoute.MOVIE}/${id}`);
    }}
    onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave}>
      <video src={previewSrc} poster={poster} width="280" height="175" ref={video}/>
    </div>
    <h3 className="small-movie-card__title">
      <Link to={`${AppRoute.MOVIE}/${id}`} className="small-movie-card__link">{title}</Link>
    </h3>
  </article>;
};

MovieCardSmall.propTypes = {
  movie: movieDetails,
  onCardClick: PropTypes.func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  onCardClick(route) {
    dispatch(redirectToRoute(route));
  },
});

export {MovieCardSmall};
export default connect(null, mapDispatchToProps)(MovieCardSmall);
