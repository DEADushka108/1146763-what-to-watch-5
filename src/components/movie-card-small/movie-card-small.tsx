import React, {useCallback, useEffect, useRef, useState, ReactNode} from 'react';
import {Link} from 'react-router-dom';
import {AppRoute} from '../../utils/const';
import {redirectToRoute} from '../../store/redirect/redirect.js';
import {connect} from 'react-redux';
import {Movie} from '../../types/movie';

interface Props {
  movie: Movie;
  children: ReactNode | ReactNode[];
  onCardClick: (route: string) => void;
}

const VIDEO_PLAY_TIMEOUT = 1000;

const MovieCardSmall = (props: Props) => {
  const {movie, onCardClick} = props;
  const {id, title, poster, previewSrc} = movie;
  const video = useRef<HTMLVideoElement>();
  const [isPlay, setPlayStatus] = useState(false);
  let timeout;

  const handleMouseEnter = useCallback(() => {
    setPlayStatus(true);
    timeout = setTimeout(() => {
      if (!isPlay && video.current) {
        video.current.muted = true;
        video.current.play();
      }
    }, VIDEO_PLAY_TIMEOUT);
  }, [video]);

  const handleMouseLeave = useCallback(() => {
    setPlayStatus(false);
    clearTimeout(timeout);
    video.current.load();
  }, [video]);

  useEffect(() => {
    if (!video.current) {
      setPlayStatus(false);
      return;
    }
    return;
  }, [video.current]);

  return <article className="small-movie-card catalog__movies-card" >
    <div className="small-movie-card__image" onClick={() => {
      onCardClick(`${AppRoute.MOVIE}/${id}`);
    }}
    onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave}>
      <video src={previewSrc} poster={poster} width="280" height="175" ref={video}/>
    </div>
    <h3 className="small-movie-card__title">
      <Link to={`${AppRoute.MOVIE}/${id}`} className="small-movie-card__link">{title}</Link>
    </h3>
  </article>;
};

const mapDispatchToProps = (dispatch) => ({
  onCardClick(route) {
    dispatch(redirectToRoute(route));
  },
});

export {MovieCardSmall};
export default connect(null, mapDispatchToProps)(MovieCardSmall);
