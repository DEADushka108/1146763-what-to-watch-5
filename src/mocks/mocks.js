export * from './const/const.js';
export * from './movies/movies.js';
export * from './reviews/reviews.js';
export * from './store/store.js';
export * from './user/user.js';
