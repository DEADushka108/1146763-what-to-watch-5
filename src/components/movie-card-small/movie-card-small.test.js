import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';
import {featuredMovie, noop} from '../../mocks/mocks.js';
import {MovieCardSmall} from './movie-card-small.jsx';

it(`Should MovieCardSmall component render correctly`, () => {
  const tree = renderer
    .create(
        <MemoryRouter>
          <MovieCardSmall
            movie={featuredMovie}
            onCardClick={noop}
          >
          </MovieCardSmall>
        </MemoryRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
