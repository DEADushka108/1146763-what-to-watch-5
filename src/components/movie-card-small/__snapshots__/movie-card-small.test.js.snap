// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should MovieCardSmall component render correctly 1`] = `
<article
  className="small-movie-card catalog__movies-card"
>
  <div
    className="small-movie-card__image"
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div />
  </div>
  <h3
    className="small-movie-card__title"
  >
    <a
      className="small-movie-card__link"
      href="/films/1"
      onClick={[Function]}
    >
      Macbeth
    </a>
  </h3>
</article>
`;
